src/cr/cube/crunch_cube.py

```diff
@@ -200,82 +200,189 @@ def _as_array(self, include_missing=False, get_non_selected=False,
-        inserted_inds = self.inserted_hs_indices() if transforms else []
-        inserted_cols = np.array(inserted_inds[1] if len(inserted_inds) > 1 else [])
-        axis = 0 if len(self.dimensions) < 3 else 1
-        col_margin = self.margin(include_transforms_for_dims=transforms, axis=axis)
-
-        # don't consider insertions outside the scope of the columns
-        inserted_cols = inserted_cols[inserted_cols < len(col_margin)]
-
-        ind_inserted = np.zeros(col_margin.shape, dtype=bool)
-        if any(inserted_cols):
-            ind_inserted[inserted_cols] = True
-        col_prune = np.logical_or(col_margin == 0, np.isnan(col_margin))
-        col_prune = np.logical_and(col_prune, ~ind_inserted)
-
-        return col_prune
+    def _prune_3d_indices(self, transforms=None):
+        row_margin = self.margin(
+            include_transforms_for_dims=transforms,
+            axis=self.row_direction_axis
+        )
+        col_margin = self.margin(
+            include_transforms_for_dims=transforms,
+            axis=self.col_direction_axis,
+        )
+        row_inserted_indices = (
+            self.inserted_rows_inds if transforms else []
+        )
+        col_inserted_indices = (
+            self.inserted_col_inds if transforms else []
+        )
+        prune_indices = []
+
+        # It's necessary to go with a for loop, since we don't know if it's
+        # going to be row or a colunm (and across which axis we might
+        # need to sum, in case of 2D margins).
+        for rm, cm in zip(row_margin, col_margin):
+            if rm.ndim > 1:
+                rm = np.sum(rm, axis=1)
+            if cm.ndim > 1:
+                cm = np.sum(cm, axis=0)
+
+            prune_indices.append((
+                self._margin_pruned_indices(rm, row_inserted_indices),
+                self._margin_pruned_indices(cm, col_inserted_indices),
+            ))
+        return prune_indices
+
+    @property
+    def row_direction_axis(self):
+        # when dealing with 1D MR cubes, axis for margin should be 0
+        if len(self.dimensions) == 1 and self.has_mr:
+            return 0
+        elif len(self.dimensions) == 3 and not self.is_double_mr:
+            return 2
+        return 1
+
+    @property
+    def inserted_rows_inds(self):
+        inserted_inds = self.inserted_hs_indices()
+        row_dim_ind = 0 if len(self.dimensions) < 3 else 1
+        return np.array(
+            inserted_inds[row_dim_ind] if len(inserted_inds) else []
+        )
+
+    @staticmethod
+    def _margin_pruned_indices(margin, insertions):
+        ind_inserted = np.zeros(margin.shape, dtype=bool)
+        if any(insertions):
+            ind_inserted[insertions] = True
+        pruned_ind = np.logical_or(margin == 0, np.isnan(margin))
+        pruned_ind = np.logical_and(pruned_ind, ~ind_inserted)
+
+        return pruned_ind
+
+    @property
+    def col_direction_axis(self):
+        return 0 if len(self.dimensions) < 3 else 1
```

Comment:
Is this the best way to get at this? Shouldn't it be something more like `len(self.dimensions)-2` or possibly to just always be `-2` (I'm honestly not sure if you can negative index here). 

But it's my understanding that with numpy, the columns is always second to last. That is, for 3d its: [x, C, R] and 4d its: [x, x, C, R] and 5d its: [x, x, x, C, R] and 6d its: [x, x, x, x, C, R]